ctk/common/src/main/kotlin/org/codice/compliance/utils/XMLDecryptor.kt

```diff
@@ -0,0 +1,163 @@
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package org.codice.compliance.utils
+
+import org.apache.wss4j.common.crypto.JasyptPasswordEncryptor
+import org.apache.wss4j.common.crypto.Merlin
+import org.apache.xml.security.encryption.XMLCipher
+import org.bouncycastle.jce.provider.BouncyCastleProvider
+import org.codice.compliance.allChildren
+import org.codice.compliance.children
+import org.w3c.dom.Document
+import org.w3c.dom.Element
+import org.w3c.dom.Node
+import java.io.File
+import java.security.Key
+import java.security.Security
+import java.util.Properties
+
+class XMLDecryptor {
+    companion object {
+        // One time initialization
+        init {
+            Security.addProvider(BouncyCastleProvider())
+            org.apache.xml.security.Init.init()
+        }
+
+        private const val ENCRYPTION_FILE_NAME = "encryption.properties"
+        private const val BC_PROVIDER = "BC"
+
+        private val serverPrivateKey by lazy {
+
+            val encryptionFile = File(ClassLoader.getSystemResource(ENCRYPTION_FILE_NAME).toURI())
+            checkNotNull(encryptionFile)
+
+            val encryptionProperties = encryptionFile.inputStream().use {
+                Properties().apply {
+                    load(it)
+                }
+            }
+
+            val keystorePassword = encryptionProperties
+                    .getProperty("org.apache.ws.security.crypto.merlin.keystore.password")
+            val keystorePrivateKeyPassword = encryptionProperties
+                    .getProperty("org.apache.ws.security.crypto.merlin.keystore.private.password")
+            val keystorePrivateKeyAlias = encryptionProperties
+                    .getProperty("org.apache.ws.security.crypto.merlin.keystore.alias")
+
+            // Using Apache Merlin to read the keystore
+            val merlin = Merlin(encryptionProperties,
+                    ClassLoader.getSystemClassLoader(),
+                    JasyptPasswordEncryptor(keystorePassword))
+
+            merlin.getPrivateKey(keystorePrivateKeyAlias, keystorePrivateKeyPassword)
+        }
+
+        /**
+         * Takes in an EncryptedID, EncryptedAssertion, or EncryptedAttribute element and returns
+         * the decrypted node.
+         *
+         * Only supports up to one level encryption using the EncryptedKey element. In other words,
+         * the data must be directly encrypted by the CTK's public key, or, if another key is used
+         * to encrypt the data, that key must be encrypted using the CTK's public key and put into
+         * the EncryptedData element.
+         *
+         * @param node An encrypted {@code Node} to be decrypted
+         * @return Node The decrypted {@code Node}
+         */
+        @Suppress("TooGenericExceptionCaught"
+                /* Decryption error handling is independent of Exception type */)
+        fun decryptAndReplaceNode(node: Node): Node {
```

Comment:
✏️ I'd recommend either making this function `internal` so that only files in `ctk/common` can access it OR having a require statement requiring that `node.localName` is equal to one of the 3 possible encrypted elements. Maybe do both.

Reply:
The javadoc mentions the allowable encrypted elements.